c/CMakeLists.txt

```diff
@@ -204,6 +239,148 @@ else()
   message(FATAL_ERROR "BLAKE3_SIMD_TYPE is set to an unknown value: '${BLAKE3_SIMD_TYPE}'")
 endif()
 
+if(BLAKE3_NO_AVX2)
+  target_compile_definitions(blake3 PRIVATE BLAKE3_NO_AVX2)
+endif()
+if(BLAKE3_NO_AVX512)
+  target_compile_definitions(blake3 PRIVATE BLAKE3_NO_AVX512)
+endif()
+if(BLAKE3_NO_SSE2)
+  target_compile_definitions(blake3 PRIVATE BLAKE3_NO_SSE2)
+endif()
+if(BLAKE3_NO_SSE41)
+  target_compile_definitions(blake3 PRIVATE BLAKE3_NO_SSE41)
+endif()
+
+if(NOT WIN32)
+  add_executable(example
+    example.c)
+  target_link_libraries(example PRIVATE blake3)
+endif()
+
+if(WIN32 AND CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+  message(WARNING "BLAKE3_TBB_PARALLELISM requires building with Clang or MSVC on Windows")
+  set(BLAKE3_TBB_PARALLELISM OFF)
+endif()
+if(BLAKE3_TBB_PARALLELISM)
+  find_package(TBB 2021.11.0 QUIET)
+  if(NOT TBB_FOUND)
+    set(BLAKE3_TBB_PARALLELISM OFF)
+  else()
+    target_sources(blake3
+      PRIVATE
+        blake3_tbb.cpp)
+    target_link_libraries(blake3
+      PUBLIC
+        # Make shared TBB a transitive dependency. The consuming program is technically not required
+        # to link TBB in order for libblake3 to function but we do this in order to prevent the
+        # possibility of multiple separate TBB runtimes being linked into a final program in case
+        # the consuming program also happens to already use TBB.
+        TBB::tbb)
+    target_compile_definitions(blake3
+      PUBLIC
+        BLAKE3_TBB_PARALLELISM)
+    target_compile_options(blake3
+      PRIVATE
+        $<$<COMPILE_LANGUAGE:CXX>:
+          $<$<CXX_COMPILER_ID:Clang>:-fno-exceptions -fno-rtti>
+          $<$<CXX_COMPILER_ID:GNU>:-fno-exceptions -fno-rtti>
+          $<$<CXX_COMPILER_ID:MSVC>:/D_HAS_EXCEPTIONS=0 /GR>
+        >)
+  endif()
+endif()
+
+if(BLAKE3_TESTING)
+  get_target_property(BLAKE3_SOURCES blake3 SOURCES)
+  get_target_property(BLAKE3_COMPILE_DEFINITIONS blake3 COMPILE_DEFINITIONS)
+  get_target_property(BLAKE3_COMPILE_OPTIONS blake3 COMPILE_OPTIONS)
+  get_target_property(BLAKE3_LINK_LIBRARIES blake3 LINK_LIBRARIES)
+  get_target_property(BLAKE3_LINK_OPTIONS blake3 LINK_OPTIONS)
+
+  # test library target
+  add_library(blake3-testing ${BLAKE3_SOURCES})
+  add_library(BLAKE3::blake3-testing ALIAS blake3-testing)
+  target_compile_definitions(blake3-testing PUBLIC BLAKE3_TESTING)
+  if(BLAKE3_COMPILE_DEFINITIONS)
+    target_compile_definitions(blake3-testing PUBLIC
+      ${BLAKE3_COMPILE_DEFINITIONS})
+  endif()
+  if(BLAKE3_COMPILE_OPTIONS)
+    target_compile_options(blake3-testing PRIVATE
+      ${BLAKE3_COMPILE_OPTIONS}
+      -O3
+      -Wall
+      -Wextra
+      -pedantic
+      -fstack-protector-strong
+      -D_FORTIFY_SOURCE=2
+      -fPIE
+      -fvisibility=hidden
+      -fsanitize=address,undefined
```

Comment:
While these are good flags for testing _in principle_, they belong into a toolchain file. We should always test with the flags provided by the user. The test program should just link against the `blake3` library target.